### ISSUE TITLE
Can not save model using model.save following multi_gpu_model

### ISSUE THREAD

**Issue by SimonWalsh1000** [fchollet/keras#8446]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [ ] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [ ] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by SimonWalsh1000** [comment]
This is a short and simple issue. Following upgrading to Keras 2.0.9 I have been using the multi_gpu_model utility but I can seem to save my models or best weights using model.save. 

The error I get is 

`TypeError: can’t pickle module objects when using model.save following this multi-gpu implementation`

I suspect there is some problem gaining access to the model object. Is there a work around this issue?

**Comment by datumbox** [comment]
Are you trying to save directly the model which is parallelized? That can be the problem. Could you provide a snippet that shows the steps you are following?

**Comment by SimonWalsh1000** [comment]
Yes - thats exactly what I am doing - I'm building the model normally, then calling the multi gpu model method function on it and after training, trying to save it with model.save. Does this need to be converted back to a non-parallelized version of the model? If so, how is this achieved?

**Comment by fchollet** [comment]
For now we recommend saving the original (template) model instead of the parallel model. I.e. call `save` on the model you passed to `multi_gpu_model`, not the model returned by it.

Both models share the same weights.

**Comment by SimonWalsh1000** [comment]
That's fantastic, and straightforward advice. 

Best
Simon


SLFWalsh MD MRCP FFRRCSI
Consultant Radiologist
Kings College Hospital Foundation Trust

**Comment by SimonWalsh1000** [comment]
Just to clarify - do you mean call model.save or model.save_weights on the template model at the end of training?

**Comment by fchollet** [comment]
Yes. Either method should work fine.

**Comment by DingkunLiu** [comment]
But how can I save the optimizer state by just saving the template model?

**Comment by s-zk** [comment]
I am on keras 2.1.2 and encountered same problem. It has been solved by follow [this answer](https://stackoverflow.com/questions/47210811/can-not-save-model-using-model-save-following-multi-gpu-model-in-keras#answer-48066771) on StackOverflow. Hope it helps.

**Comment by taalbrecht** [comment]
> Closing as this is resolved

Has this been resolved by a commit and can models and weights be saved as expected when using multi_gpu_model?

**Comment by Heisenberg0391** [comment]
So, as mentioned above, i should train with parallel_model but save the origin model. But what if i want save weights on every epoch as checkpoints using a callback, what should i do?

**Comment by jashshah** [comment]
Are there any updates regarding this either on keras or on tf.keras?
